src/flamenco/vm/syscall/fd_vm_syscall_cpi_common.c

```diff
@@ -315,54 +382,170 @@ TODO: error codes
+    ulong region_idx = vm->acc_region_metas[ instr_acc_idx ].region_idx;
+    uint original_len = vm->acc_region_metas[ instr_acc_idx ].has_data_region ? 
+                        vm->input_mem_regions[ region_idx ].region_sz : 0U;
+
+    int zero_all_mapped_spare_capacity = 0;
+    /* This case can only be triggered if the original length is more than 0 */
+    if( callee_acc_rec->const_meta->dlen < original_len ) {
+      ulong new_len = callee_acc_rec->const_meta->dlen;
+      /* Allocate into the buffer to make sure that the original data len
+         is still valid but don't change the dlen. Zero out the rest of the
+         memory which is not used. */
+      fd_instr_borrowed_account_modify( vm->instr_ctx, pubkey, original_len, &callee_acc_rec );
+      callee_acc_rec->meta->dlen = new_len;
+      zero_all_mapped_spare_capacity = 1;
+    }
+
+    /* Update the account data region if an account data region exists. We
+       know that one exists iff the original len was non-zero. */
+    ulong acc_region_idx = vm->acc_region_metas[instr_acc_idx].region_idx;
+    if( original_len && vm->input_mem_regions[ acc_region_idx ].haddr!=(ulong)callee_acc_rec->data ) {
+      vm->input_mem_regions[ acc_region_idx ].haddr = (ulong)callee_acc_rec->data;
+      zero_all_mapped_spare_capacity = 1;
+    }
+
+    ulong prev_len = caller_acc_metadata_len;
+    ulong post_len = callee_acc_rec->const_meta->dlen;
+
+    /* Do additional handling in the case where the data size has changed in
+       the course of the callee's CPI. */
+    if( prev_len!=post_len ) {
+      /* There is an illegal data overflow if the post len is greater than the
+         original data len + the max resizing limit (10KiB). Can't resize the
+         account if the deprecated loader is being used */
+      ulong max_increase = vm->is_deprecated ? 0UL : 10240UL;
+      if( FD_UNLIKELY( post_len>fd_ulong_sat_add( original_len, max_increase ) ) ) {
+        return FD_EXECUTOR_INSTR_ERR_INVALID_REALLOC;
+      }
+      /* There is additonal handling in the case where the account is larger
+         than it was previously, but it has still grown since it was initially
+         serialized. To handle this, we need to just zero out the now unused
+         space in the account resizing region. */
+      if( post_len<prev_len && prev_len>original_len ) {
+        ulong dirty_realloc_start = fd_ulong_max( post_len, original_len );
+        ulong dirty_realloc_len   = fd_ulong_sat_sub( prev_len, dirty_realloc_start );
+        /* We don't have to worry about multiple region writes here since we
+           know the amount to zero out is located in the account's data
+           resizing region. We intentionally write to the pointer despite
+           loading it in because we assume that the permissions were changed
+           in the callee. */
+        uchar * dirty_region = (uchar*)FD_VM_MEM_HADDR_LD( vm, caller_acc_metadata_vm_addr + dirty_realloc_start,
```

Comment:
This should be `FD_VM_MEM_HADDR_ST`

Reply:
Agave converts it to a non-writable region then converts it back. It doesn't make sense to do this if we can just do a load to get the raw pointer

Reply:
I don't think they do - they convert it to a _writable_ region to zero it out, then set it back to the same permissions it had before: https://github.com/anza-xyz/agave/blob/36323b6dcd3e29e4d6fe6d73d716a3f33927148b/programs/bpf_loader/src/syscalls/cpi.rs#L1399-L1400

In our case, we are performing a write through a pointer obtained from `FD_VM_MEM_HADDR_LD`, which is an API violation. We should use `FD_VM_MEM_HADDR_ST` to obtain the pointer, as we are performing a write through it.

I'm not suggesting we change the regions metadata like they do, just that we use our memory translation APIs correctly.

Reply:
This still feels off to me, we are mis-using the API. I wonder if we should create a new `_ST_UNCHECKED` macro or something to make the semantics clear. Just worried someone will make this stricter in the future and that will break things.

Reply:
adding a st_write_unchecked